torch/_dynamo/variables/higher_order_ops.py

```diff
@@ -538,6 +538,58 @@ def speculate_subgraph(
                 graph.lint()
                 lifted_freevars = subtracer.lifted_freevars
 
+                # NOTE: [HigherOrderOperator subgraph input ordering]
+                # The input ordering of the higher order ops is determined by the order of
+                # the creatation of the placehoder.
+                # Mannually created inputs are created in validate_args_and_maybe_create_graph_inputs before
+                # speculating subgraph.
+                # During subgraph speculation, we may lift closured tensors and free symbols as inputs,
+                # their ordering is determined by the time they are lifted: earlier lifted ones precede later
+                # lifted ones.
+                #
+                # Suppose the placeholders are
+                # O1, O2, X1, O3, O4, X2, X3, O5 where Xs are lifted phs
+                # The following code re-order the placeholders to
+                # O1, O2, O3, O4, O5, X1, X2, X3
+                def move_lifted_freevars_phs_to_end(
```

Comment:
How does Dynamo and AOTDispatcher handle their calling convention? I'm wondering if there are concrete reasons why we should one calling convention over the other.

Reply:
Is this function necessary because we return all the freevars at the end?

Reply:
yeah, it's necessary because manually created inputs and lifted_freevars are treated differently in hops. The previous contract has been manually created ones are always before lifted ones. With auto lifting, we could have cases like users manually create_graph_input for a tensor t, we auto-lift the symbols in it immediately, which causes the phs change to s1, s2, t. This pass move s1, s2 to be after all the manual created ones and the signature becomes t, s1, s2. If a hop can handle llifted freevars previously, it can automatically handle lifted free symbols in dynamo.

Reply:
Reordering is kind of janky, is it possible to change the current fx node insertion point after we add each input?